src/linear_algebra/matrix/circulant.lean

```diff
@@ -0,0 +1,238 @@
+/-
+Copyright (c) 2021 Lu-Ming Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Lu-Ming Zhang
+-/
+import linear_algebra.matrix.symmetric
+import data.polynomial.monomial
+import data.matrix.pequiv
+import data.equiv.fin
+
+/-!
+# Circulant matrices
+
+This file contains the definition and basic results about circulant matrices.
+
+## Main results
+
+- `matrix.cir`: introduce the definition of a circulant matrix
+                generated by a given vector `v : I → α`.
+
+## Implementation notes
+
+`fin.foo` is the `fin n` version of `foo`.
+Namely, the index type of the circulant matrices in discussion is `fin n`.
+
+## Tags
+
+cir, matrix
+-/
+
+variables {α β I J R : Type*} {n : ℕ}
+
+namespace matrix
+open_locale matrix big_operators
+
+/-- Given the condition `[has_sub I]` and a vector `v : I → α`,
+    we define `cir v` to be the circulant matrix generated by `v` of type `matrix I I α`. -/
+def cir [has_sub I] (v : I → α) : matrix I I α
```

Comment:
I'd be tempted to call this the full name,
```suggestion
def circulant [has_sub I] (v : I → α) : matrix I I α
```

Reply:
Do I also change `cir` to `circulant` in all the other places? I used `circulant` at the beginning in fact, but then I changed to `cir` as I found `circulant` is too wordy and I feel `cir` is clear here.

Reply:
Yes, if you change one you should change them all.`circulant` isn't much longer than `transpose` or `diagonal`, and is used far less; so I think the longer but clearer name is better.